### PR TITLE
SignUp: Improves a11y for the select domain step.

### LINKED ISSUE #21658: Signup: Screen reader issues with domain picker step
Testing signup with VoiceOver + Safari.

Once I land on the domain picker step in signup, I'm autofocused into the search bar, so I don't hear the context of the page, or what I'm supposed to be searching for.

I think this can be fixed by adding `aria-describedby`, referencing the element with some description text. I'd also suggest not autofocusing via JS, but instead just using the html input attribute `autofocus`. [More thoughts on autofocus](http://www.brucelawson.co.uk/2009/the-accessibility-of-html-5-autofocus/)

Additionally, there was no feedback that any results had populated. Perhaps we could add a section with [`aria-live`](https://webaim.org/techniques/aria/#dynamic), to update with something like "X results found"? This would need to be debounced, so it's not constantly interrupting itself.

### PR DESCRIPTION
**Pull request by rodrigoi**
As suggested on #21658, this PR makes the following changes to improve a11y support on the _Domains_ step of the sign up process:

- `Search` now uses `autoFocus` instead of JS, and adds `aria-describedBy` support.
- Domain search results now is defined as a `aria-live` region.

Also, `FormattedHeader` now has a `id` property and there are some minor updates on the `Search` component.

How to Test:

- navigate to `/start/`
- activate the screen reader of choice.
- click `Continue`, to access the _Domains_ step.

The _Domains_ step should load with the focus on the search bar, and the screen reader should announce the search results automatically when you finish typing.

Since this makes changes on the `Search` component, there should be no regressions on components that depend on it. Components like the `SiteSelector` and the `SectionNav` should work as expected.

Fixes #21658

### CODE REVIEW

File: client/components/search/index.jsx [hunk 1]
```diff
@@ -345,36 +325,35 @@ class Search extends Component {
...
 				>
 					{ ! this.props.hideOpenIcon && <Gridicon icon="search" className="search__open-icon" /> }
 				</div>
 				<div className={ fadeDivClass }>
 					<input
 						type="search"
-						id={ 'search-component-' + this.state.instanceId }
+						id={ 'search-component-' + this.instanceId }
+						autoFocus={ this.props.autoFocus } // eslint-disable-line jsx-a11y/no-autofocus
+						aria-describedBy={ this.props.describedBy }
```
Comment (vindl) [inline client/components/search/index.jsx]: Looks like there is a typo here: `aria-describedby`.

Comment (rodrigoi) [reply client/components/search/index.jsx]: that's just _React_ beign silly. It get's rendered correctly:
screen shot 2018-03-16 at 14 41 28

**Comment by gwwar** [comment]
> FAIL  client/components/language-picker/test/modal.js (5.507s)

Looks like we need to update a test, or rebase.

File: client/components/search/index.jsx (hunk 1 above)
Comment (rodrigoi) [reply client/components/search/index.jsx]: In fact, you are right. From the [docs](https://reactjs.org/docs/accessibility.html):

> Whereas most DOM properties and attributes in React are camelCased, these attributes should be lowercased.

Weird thing is that _React_ does not complain. Updated on 312f6d02debfb5a94ae139f710b855e2280eaf1f.

**Comment by kwight** [comment]
I fired up VoiceOver+Chrome, worked as expected for me. I'll leave the final 👍 to someone more knowledgeable :)

File: client/components/domains/domain-search-results/index.jsx [hunk 2]
```diff
@@ -245,7 +245,7 @@ class DomainSearchResults extends React.Component {
 
 	render() {
 		return (
-			<div className="domain-search-results">
+			<div className="domain-search-results" aria-live="polite" aria-relevant="additions">
```
Comment (ryelle) [inline client/components/domains/domain-search-results/index.jsx]: Technically this works, but there's a slight delay between the paid domains & the "free" option, which means that the screen reader is in the process of reading the first paid result when the free result appears, and so it ends up interrupting itself to read the free row, then stops.

Another solution to this might be something like this, which would allow for live feedback without having to read the entire table to the screen reader.

```
<div aria-live="polite">
    <ScreenReaderText>5 domains found</ScreenReaderText>
</div>
```

Comment (rodrigoi) [reply client/components/domains/domain-search-results/index.jsx]: 👍

Comment (rodrigoi) [reply client/components/domains/domain-search-results/index.jsx]: and any way to create a link between that counter and the actual search results?

Comment (ryelle) [reply client/components/domains/domain-search-results/index.jsx]: I'm not sure what you mean– the counter would basically be an audio "notice", the user likely wouldn't intentionally navigate focus to it. once they've typed out their search, they would navigate into the search results.

Comment (gwwar) [reply client/components/domains/domain-search-results/index.jsx]: > and any way to create a link between that counter and the actual search results

If you mean how to get the count, the component should have access to that. I suppose it'd be something like domain availability + domain suggestions (when they've loaded)

File: client/components/search/index.jsx [hunk 3]
```diff
@@ -93,16 +92,17 @@ class Search extends Component {
...
-		hasFocus: false,
-	};
+	constructor( props ) {
+		super( props );
 
-	componentWillMount() {
-		this.setState( {
-			instanceId: ++Search.instances,
-		} );
+		instanceCount++;
```
Comment (gwwar) [inline client/components/search/index.jsx]: Careful when refactoring base components, this opens up the regression testing needed for this PR. 

It's usually easier to pull out changes like this to it's own PR so we can clearly explain why it's needed and verify that other search instances have not regressed.

File: client/components/search/index.jsx [hunk 4]
```diff
@@ -345,36 +325,35 @@ class Search extends Component {
 				<Spinner />
 				<div
 					className="search__icon-navigation"
-					ref="openIcon"
+					ref={ openIcon => ( this.openIcon = openIcon ) } // eslint-disable-line react/jsx-no-bind
```
Comment (gwwar) [inline client/components/search/index.jsx]: Another option would be to pass a short function ;)

File: client/components/search/index.jsx [hunk 5]
```diff
@@ -345,36 +325,35 @@ class Search extends Component {
...
 						className={ inputClass }
 						placeholder={ placeholder }
 						role="search"
 						value={ searchValue }
-						ref="searchInput"
-						onInput={
-							this.onChange
-							/* onChange has bug IE11 React15 https://github.com/facebook/react/issues/7027 */
-						}
+						ref={ input => ( this.searchInput = input ) } //eslint-disable-line react/jsx-no-bind
```
Comment (gwwar) [inline client/components/search/index.jsx]: Ditto

File: client/components/search/index.jsx [hunk 6]
```diff
@@ -389,7 +368,10 @@ class Search extends Component {
 
 	renderStylingDiv = () => {
 		return (
-			<div className="search__text-overlay" ref="overlay">
+			<div
+				className="search__text-overlay"
+				ref={ overlay => ( this.overlay = overlay ) } //eslint-disable-line react/jsx-no-bind
```
Comment (gwwar) [inline client/components/search/index.jsx]: So many refs 😄

**Comment by gwwar** [comment]
So overall, I think this is a good improvement! One thing probably I'd suggest that may be out of the scope for this, is that I noticed we don't describe which domain suggestion it is when "select" is focused.

- [ ] Let's get the Search changes pulled out into it's own PR 
- [ ] We can run the full suite against it (use the `Needs e2e testing`) label, and ask flow for help manually testing the other search instances. I'll also help get this out quickly.
- [ ] Rebase and land this PR if folks didn't have other blockers.

File: client/components/domains/domain-search-results/index.jsx (hunk 2 above)
Comment (rodrigoi) [reply client/components/domains/domain-search-results/index.jsx]: I've updated the code on 7a4fe27f76fba7a3ea7060f0e7e69073ee093320 with `ScreenReaderText`. Now there are two `aria-live` regions:

- Domain Availability / Result or TLD not available.
- Result count.

I think I've covered all the cases (with or without suggested domains, weird TLDs, owned domains, etc), and I think that the voiceover makes sense.
The only problem I think we have is that we always show 10 domain suggestions for the signup flows, and for the domain add flow on _My Sites_ we initially show 2, and then 10.
The suggestions could have a WordPress.com domain or not (like on _My Sites / Domains_), but it's always 10.
Does this matter? Should we include a more descriptive text?

Comment (gwwar) [reply client/components/domains/domain-search-results/index.jsx]: Is it super important that the count is correct? Could we maybe phrase text to say, something along the lines of domain results found, tab to examine/review results?

**Comment by benhuberman** [comment]
Hi @rodrigoi -- I just read through this conversation and don't see any copy to review; could it be I'm missing something or looking in the wrong place?

File: client/components/domains/domain-search-results/index.jsx (hunk 2 above)
Comment (ryelle) [reply client/components/domains/domain-search-results/index.jsx]: No, the count doesn't really matter, the point is to have some kind of audio feedback - so that would work.

Comment (gwwar) [reply client/components/domains/domain-search-results/index.jsx]: Apologies, @benhuberman we'll try and remember to add a direct comment for y'all when adding the `needs copy review` label. In this case I think we need some help with text that lets folks know that a search result has completed, without needing to describe an exact count. Eg 

```jsx
<div aria-live="polite">
    <ScreenReaderText>Domains found! Tab to select an option. </ScreenReaderText>
</div>
```

File: client/components/domains/domain-search-results/index.jsx [hunk 7]
```diff
@@ -175,7 +176,11 @@ class DomainSearchResults extends React.Component {
 		}
 
 		return (
-			<div className="domain-search-results__domain-availability">
+			<div
+				className="domain-search-results__domain-availability"
+				aria-live="polite"
```
Comment (ryelle) [inline client/components/domains/domain-search-results/index.jsx]: This doesn't seem to do anything in my VoiceOver testing – I think because this entire element is added/changed at once, rather than changing content inside of it.

File: client/components/domains/domain-search-results/index.jsx (hunk 2 above)
Comment (benhuberman) [reply client/components/domains/domain-search-results/index.jsx]: Thanks @gwwar! It's sometimes hard to propose copy without more context about the flow or a screenshot that shows how it's displayed. In this case, though, it seems like a fairly minor / straightforward case... how about:
> Here are some domain ideas -- tap to select one of the options.

Let me know if this doesn't quite fit (space-wise or otherwise)!

Comment (ryelle) [reply client/components/domains/domain-search-results/index.jsx]: @benhuberman I'm not sure that works for this case, here's some more context: This text won't be shown visually- it'll be read aloud to screen reader users when the domain results change, so that when they're typing in the search box, there is audio feedback that the results have updated. Then they can use the keyboard to get to the results.

File: client/components/domains/domain-search-results/index.jsx (hunk 7 above)
Comment (rodrigoi) [reply client/components/domains/domain-search-results/index.jsx]: I've removed this for now on 4a95bfb315198bfecd62eabfac3526e36e9a49a8 because it doesn't work entirely as expected. To make it work, I'll have to reorder some elements and change the way some of the components render so they update content as `aria-live` expects. I would like to follow that up on a new PR so this don't became stale.

**Comment by rodrigoi** [comment]
thanks @kwight @gwwar @vindl @ryelle and @benhuberman for taking the time to review this PR. I think it's ready for another check, since #23600 has been merged without regressions.

I've created a new issue for the availability warnings (#23649) so we can move this PR forward.

**Comment by kwight** [comment]
I gave this a run-through again with VoiceOver. The focus seems fine but this part:

> and the screen reader should announce the search results automatically when you finish typing

seemed odd. Hitting `return` after entering a string made VoiceOver read the `formatted-header` part again, and then it got to the search results; is this expected behaviour?

File: client/components/search/index.jsx [hunk 8]
```diff
@@ -349,6 +346,10 @@ class Search extends Component {
 					<input
 						type="search"
 						id={ 'search-component-' + this.instanceId }
+						autoFocus={ this.props.autoFocus } // eslint-disable-line jsx-a11y/no-autofocus
```
Comment (gwwar) [inline client/components/search/index.jsx]: is this against recommendations?

File: client/components/formatted-header/index.jsx [hunk 9]
```diff
@@ -19,7 +19,7 @@ function FormattedHeader( { headerText, subHeaderText } ) {
 	} );
 
 	return (
-		<header className={ classes }>
+		<header id={ 'formatted-header' } className={ classes }>
```
Comment (gwwar) [inline client/components/formatted-header/index.jsx]: Would we never have more than one formatted-header on a page?

Comment (gwwar) [reply client/components/formatted-header/index.jsx]: Feels like we can pass this in as a prop

File: client/components/search/index.jsx (hunk 8 above)
Comment (rodrigoi) [reply client/components/search/index.jsx]: This triggers a warning, but it's the suggested solution on https://github.com/Automattic/wp-calypso/issues/21658.

The problem is that `autoFocus` could prevent screen readers to read the content before the component, which in this case we are using `aria-describedby` to reference, so it's announced correctly by screen readers.

File: client/components/formatted-header/index.jsx (hunk 9 above)
Comment (rodrigoi) [reply client/components/formatted-header/index.jsx]: We don't have any cases of more than one `FormattedHeader` on a page, but it's not future-proof. Updated on ec1f236b33aef21433712a6afa69816fc7d6077d to get it as a `prop` from the `StepWrapper`.

File: client/components/search/index.jsx (hunk 8 above)
Comment (gwwar) [reply client/components/search/index.jsx]: Let's add another comment explaining this. Folks might try to remove this otherwise.

**Comment by klimeryk** [comment]
Also cc @jsnmoon who is implementing most (all?) of the frontend 👍
